File: android/core/src/main/kotlin/com/iblocker/core/log/QueryLogRecord.kt
package com.iblocker.core.log

enum class LogVerdict(val raw: Int) {
    ALLOWED(0),
    BLOCKED(1),
    FAILED(2),
    ;

    companion object {
        fun from(raw: Int): LogVerdict? = entries.firstOrNull { it.raw == raw }
    }
}

/**
 * One DNS decision, encoded as a fixed 64-byte record:
 *
 *     0   timestampMillis  u64 LE
 *     8   qtype            u16 LE
 *     10  verdict          u8
 *     11  listId           u8   (reserved, 0 in v1)
 *     12  domainLen        u8
 *     13  flags            u8   (reserved)
 *     14  reserved         u16
 *     16  domain UTF-8, up to 48 bytes, zero padded
 */
data class QueryLogRecord(
    val timestampMillis: Long,
    val qtype: Int,
    val verdict: LogVerdict,
    val listId: Int = 0,
    val domain: String,
) {
    fun encoded(): ByteArray {
        val out = ByteArray(RECORD_SIZE)
        for (i in 0 until 8) {
            out[i] = ((timestampMillis ushr (i * 8)) and 0xFF).toByte()
        }
        out[8] = (qtype and 0xFF).toByte()
        out[9] = ((qtype shr 8) and 0xFF).toByte()
        out[10] = verdict.raw.toByte()
        out[11] = listId.toByte()

        var domainBytes = domain.toByteArray(Charsets.UTF_8)
        if (domainBytes.size > MAX_DOMAIN_BYTES) {
            domainBytes = domainBytes.copyOfRange(domainBytes.size - MAX_DOMAIN_BYTES, domainBytes.size)
        }
        out[12] = domainBytes.size.toByte()
        domainBytes.copyInto(out, 16)
        return out
    }

    companion object {
        const val RECORD_SIZE = 64
        const val MAX_DOMAIN_BYTES = 48

        /** Decodes one 64-byte slot; returns null for torn/blank slots. */
        fun decode(bytes: ByteArray, offset: Int = 0): QueryLogRecord? {
            if (offset + RECORD_SIZE > bytes.size) return null

            var ts = 0L
            for (i in 7 downTo 0) {
                ts = (ts shl 8) or (bytes[offset + i].toLong() and 0xFF)
            }
            if (ts == 0L) return null

            val qtype = (bytes[offset + 8].toInt() and 0xFF) or ((bytes[offset + 9].toInt() and 0xFF) shl 8)
            val verdict = LogVerdict.from(bytes[offset + 10].toInt() and 0xFF) ?: return null

            val domainLength = bytes[offset + 12].toInt() and 0xFF
            if (domainLength <= 0 || domainLength > MAX_DOMAIN_BYTES) return null
            val domainBytes = bytes.copyOfRange(offset + 16, offset + 16 + domainLength)
            if (domainBytes.any { (it.toInt() and 0xFF) < 0x21 || (it.toInt() and 0xFF) > 0x7E }) return null

            return QueryLogRecord(
                timestampMillis = ts,
                qtype = qtype,
                verdict = verdict,
                listId = bytes[offset + 11].toInt() and 0xFF,
                domain = String(domainBytes, Charsets.UTF_8),
            )
        }
    }
}

File: ios/IBlockerKit/Sources/IBlockerKit/Log/QueryLogRecord.swift
import Foundation

public enum LogVerdict: UInt8, Sendable, Codable {
    case allowed = 0
    case blocked = 1
    case failed = 2
}

/// One DNS decision, encoded as a fixed 64-byte record:
///
///     0   timestampMillis  u64 LE
///     8   qtype            u16 LE
///     10  verdict          u8
///     11  listId           u8   (reserved, 0 in v1)
///     12  domainLen        u8
///     13  flags            u8   (reserved)
///     14  reserved         u16
///     16  domain UTF-8, up to 48 bytes, zero padded
public struct QueryLogRecord: Sendable, Equatable {
    public static let recordSize = 64
    public static let maxDomainBytes = 48

    public var timestampMillis: UInt64
    public var qtype: UInt16
    public var verdict: LogVerdict
    public var listId: UInt8
    public var domain: String

    public init(timestampMillis: UInt64, qtype: UInt16, verdict: LogVerdict,
                listId: UInt8 = 0, domain: String) {
        self.timestampMillis = timestampMillis
        self.qtype = qtype
        self.verdict = verdict
        self.listId = listId
        self.domain = domain
    }

    public var date: Date {
        Date(timeIntervalSince1970: TimeInterval(timestampMillis) / 1000)
    }

    public func encoded() -> [UInt8] {
        var out = [UInt8](repeating: 0, count: Self.recordSize)
        var ts = timestampMillis.littleEndian
        withUnsafeBytes(of: &ts) { for (i, b) in $0.enumerated() { out[i] = b } }
        out[8] = UInt8(qtype & 0xFF)
        out[9] = UInt8(qtype >> 8)
        out[10] = verdict.rawValue
        out[11] = listId

        var domainBytes = Array(domain.utf8)
        if domainBytes.count > Self.maxDomainBytes {
            domainBytes = Array(domainBytes.suffix(Self.maxDomainBytes))
        }
        out[12] = UInt8(domainBytes.count)
        for (i, b) in domainBytes.enumerated() { out[16 + i] = b }
        return out
    }

    /// Decodes one 64-byte slot; returns nil for torn/blank slots.
    public static func decode(_ bytes: ArraySlice<UInt8>) -> QueryLogRecord? {
        guard bytes.count == recordSize else { return nil }
        let b = Array(bytes)

        var ts: UInt64 = 0
        for i in (0..<8).reversed() { ts = ts << 8 | UInt64(b[i]) }
        guard ts != 0 else { return nil }

        let qtype = UInt16(b[8]) | UInt16(b[9]) << 8
        guard let verdict = LogVerdict(rawValue: b[10]) else { return nil }

        let domainLength = Int(b[12])
        guard domainLength > 0, domainLength <= maxDomainBytes else { return nil }
        let domainBytes = Array(b[16..<(16 + domainLength)])
        guard domainBytes.allSatisfy({ $0 >= 0x21 && $0 <= 0x7E }) else { return nil }

        return QueryLogRecord(
            timestampMillis: ts,
            qtype: qtype,
            verdict: verdict,
            listId: b[11],
            domain: String(decoding: domainBytes, as: UTF8.self)
        )
    }
}
